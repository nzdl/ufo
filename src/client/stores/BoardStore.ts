import {observable} from "mobx";



export class BoardStore {

    @observable boards = [];

    constructor(boards){
        this.boards = boards;
    }

    Add = () => {
        let board = new Board([]);
        board.boardStore = this;
        this.boards.push(board);
    }

}

export class Board {

  @observable lists = [];
  boardStore: BoardStore;

  constructor(lists){
      this.lists = lists;
  }

    Add = () => {
      let list = new List([]);
      list.board = this;
      this.lists.push(list)
    }

    Delete = (list) => {
        this.lists = this.lists.filter((c) => c !== list)
    }

}


export class List {
    @observable cards = Array<Card>();
    board: Board;

  constructor(cards){
    this.cards = cards
  }

  Add = (card) => {
      card.list = this
      this.cards.push(card)
  }

  DeleteCard = (card) => {
      this.cards = this.cards.filter((c) => c !== card)
  }
}

export class Card {
  title: string;
  constructor(title){
    this.title = title
  }

  list: List

  Delete = () => {
      this.list.DeleteCard(this)
  }
}
