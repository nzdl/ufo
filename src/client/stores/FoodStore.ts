import axios from 'axios'
import { observable } from 'mobx'

export class FoodStore {
  @observable foods = []

  async fetchFoods() {
    this.foods = await index()
  }

  async createFood({ name }) {
    await create({ name })
    await this.fetchFoods()
  }
}

async function index() {
  const response = await axios.get('/foods')
  return response.data
}

async function create({ name }) {
  return await axios.post('/foods', { name })
}

export const foodStore = new FoodStore()
