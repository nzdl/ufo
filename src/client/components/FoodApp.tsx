import { observable } from 'mobx'
import { inject, observer } from 'mobx-react'
import * as React from 'react'
import { foodStore } from '../stores/FoodStore'

@inject('foodStore')
@observer
export class FoodApp extends React.Component<any> {
  async componentDidMount() {
    await this.props.foodStore.fetchFoods()
  }

  render() {
    return <>
      <Foods foods={this.props.foodStore.foods} />
      <AddFood />
    </>
  }
}

function Foods({ foods }) {
  return <ul>{foods.map(food => <Food key={food.name} {...food} />)}</ul>
}

function Food({ name }) {
  return <li>{name}</li>
}

@inject('foodStore')
@observer
class AddFood extends React.Component<any> {
  @observable name = ''

  render() {
    const handleClick = async () => {
      await this.props.foodStore.createFood({ name: this.name })
      this.name = ''
    }

    return <>
      <input type="text" value={this.name} onChange={e => this.name = e.target.value} />
      <button onClick={handleClick}>Add</button>
    </>
  }
}
