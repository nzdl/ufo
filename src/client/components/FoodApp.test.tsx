import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { mount } from 'enzyme'
import { Provider } from 'mobx-react'
import * as React from 'react'
import { FoodStore } from '../stores/FoodStore'
import { FoodApp } from './FoodApp'

const mock = new MockAdapter(axios)

beforeEach(mock.reset)
afterAll(mock.restore)

test('FoodApp', async () => {
  const givenHamburger = () => ({ name: 'hamburger' })
  const givenFrenchFries = () => ({ name: 'french fries' })

  mock.onGet('/foods').reply(200, [givenHamburger()])

  const foodStore = new FoodStore()
  expect(foodStore.foods.length).toBe(0)

  const wrapper = mount(
    <Provider foodStore={foodStore}>
      <FoodApp />
    </Provider>
  )

  expect(wrapper).toMatchSnapshot()

  await block()
  expect(foodStore.foods.length).toBe(1)
  expect(foodStore.foods).toContainEqual(givenHamburger())
  expect(wrapper).toMatchSnapshot()

  mock.onPost('/foods', givenFrenchFries()).reply(() => {
    mock.onGet('/foods').reply(200, [givenHamburger(), givenFrenchFries()])
    return [200, givenHamburger()]
  })

  const input = wrapper.find('input')
  input.simulate('focus')
  input.simulate('change', { target: { value: 'french fries'} })

  wrapper.find('button').simulate('click')

  await block()
  expect(foodStore.foods.length).toBe(2)
  expect(foodStore.foods).toContainEqual(givenFrenchFries())
  expect(wrapper).toMatchSnapshot()
})

function block() {
  return new Promise(resolve => setTimeout(resolve, 0))
}
