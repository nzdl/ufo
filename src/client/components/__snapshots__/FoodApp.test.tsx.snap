// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodApp 1`] = `
<Provider
  foodStore={
    FoodStore {
      "foods": Array [],
    }
  }
>
  <inject-FoodApp-with-foodStore>
    <FoodApp
      foodStore={
        FoodStore {
          "foods": Array [],
        }
      }
    >
      <Foods
        foods={Array []}
      >
        <ul />
      </Foods>
      <inject-AddFood-with-foodStore>
        <AddFood
          foodStore={
            FoodStore {
              "foods": Array [],
            }
          }
        >
          <input
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            onClick={[Function]}
          >
            Add
          </button>
        </AddFood>
      </inject-AddFood-with-foodStore>
    </FoodApp>
  </inject-FoodApp-with-foodStore>
</Provider>
`;

exports[`FoodApp 2`] = `
<Provider
  foodStore={
    FoodStore {
      "foods": Array [
        Object {
          "name": "hamburger",
        },
      ],
    }
  }
>
  <inject-FoodApp-with-foodStore>
    <FoodApp
      foodStore={
        FoodStore {
          "foods": Array [
            Object {
              "name": "hamburger",
            },
          ],
        }
      }
    >
      <Foods
        foods={Array []}
      >
        <ul />
      </Foods>
      <inject-AddFood-with-foodStore>
        <AddFood
          foodStore={
            FoodStore {
              "foods": Array [
                Object {
                  "name": "hamburger",
                },
              ],
            }
          }
        >
          <input
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            onClick={[Function]}
          >
            Add
          </button>
        </AddFood>
      </inject-AddFood-with-foodStore>
    </FoodApp>
  </inject-FoodApp-with-foodStore>
</Provider>
`;

exports[`FoodApp 3`] = `
<Provider
  foodStore={
    FoodStore {
      "foods": Array [
        Object {
          "name": "hamburger",
        },
        Object {
          "name": "french fries",
        },
      ],
    }
  }
>
  <inject-FoodApp-with-foodStore>
    <FoodApp
      foodStore={
        FoodStore {
          "foods": Array [
            Object {
              "name": "hamburger",
            },
            Object {
              "name": "french fries",
            },
          ],
        }
      }
    >
      <Foods
        foods={
          Array [
            Object {
              "name": "hamburger",
            },
          ]
        }
      >
        <ul>
          <Food
            key="hamburger"
            name="hamburger"
          >
            <li>
              hamburger
            </li>
          </Food>
        </ul>
      </Foods>
      <inject-AddFood-with-foodStore>
        <AddFood
          foodStore={
            FoodStore {
              "foods": Array [
                Object {
                  "name": "hamburger",
                },
                Object {
                  "name": "french fries",
                },
              ],
            }
          }
        >
          <input
            onChange={[Function]}
            type="text"
            value="french fries"
          />
          <button
            onClick={[Function]}
          >
            Add
          </button>
        </AddFood>
      </inject-AddFood-with-foodStore>
    </FoodApp>
  </inject-FoodApp-with-foodStore>
</Provider>
`;
